tests/thread_pool.rs

```diff
@@ -0,0 +1,141 @@
+}
+
+// Test for slow joins with task burst.
+#[test]
+#[ignore]
+fn slow_join_interrupted() {
+    let thread_join_time_max = 2_000;
+    let start = Instant::now();
+
+    // Send an initial batch of million bursts.
+    let handles = (0..1_000_000)
+        .map(|_| {
+            task::blocking::spawn(async {
+                let duration = Duration::from_millis(1);
+                thread::sleep(duration);
+            })
+        })
+        .collect::<Vec<JoinHandle<()>>>();
+
+    task::block_on(join_all(handles));
+
+    // Let them join to see how it behaves under different workloads.
+    // This time join under the time window.
+    let duration = Duration::from_millis(thread_join_time_max);
+    thread::sleep(duration);
+
+    // Spawn yet another batch of work on top of it
+    let handles = (0..10_000)
+        .map(|_| {
+            task::blocking::spawn(async {
+                let duration = Duration::from_millis(100);
+                thread::sleep(duration);
+            })
+        })
+        .collect::<Vec<JoinHandle<()>>>();
+
+    task::block_on(join_all(handles));
+
+    // Slow joins shouldn't cause internal slow down
+    let elapsed = start.elapsed().as_millis() - thread_join_time_max as u128;
+    println!("Slow task join. Monotonic exec time: {:?} ns", elapsed);
+
+    // Should be less than 25_000 ns
+    // Previous implementation is around this threshold.
+    assert_eq!(elapsed < 25_000, true);
+}
+
+// This test is expensive but it proves that longhauling tasks are working in adaptive thread pool.
+// Thread pool which spawns on-demand will panic with this test.
+#[test]
+#[ignore]
+fn longhauling_task_join() {
+    let thread_join_time_max = 11_000;
+    let start = Instant::now();
+
+    // First batch of overhauling tasks
+    let _ = (0..100_000)
+        .map(|_| {
+            task::blocking::spawn(async {
+                let duration = Duration::from_millis(1000);
+                thread::sleep(duration);
+            })
+        })
+        .collect::<Vec<JoinHandle<()>>>();
+
+    // Let them join to see how it behaves under different workloads.
+    let duration = Duration::from_millis(thread_join_time_max);
+    thread::sleep(duration);
+
+    // Send yet another medium sized batch to see how it scales.
+    let handles = (0..10_000)
+        .map(|_| {
+            task::blocking::spawn(async {
+                let duration = Duration::from_millis(100);
+                thread::sleep(duration);
+            })
+        })
+        .collect::<Vec<JoinHandle<()>>>();
+
+    task::block_on(join_all(handles));
+
+    // Slow joins shouldn't cause internal slow down
+    let elapsed = start.elapsed().as_millis() - thread_join_time_max as u128;
+    println!(
+        "Long-hauling task join. Monotonic exec time: {:?} ns",
+        elapsed
+    );
+
+    // Should be less than 200_000 ns
+    // Previous implementation will panic when this test is running.
+    assert_eq!(elapsed < 200_000, true);
+}
```

Comment:
these tests are generally non-reproducible, and it's unclear if they are valuable. they tend to fail on my fairly modern 2018 i7 mbp

Reply:
when I said concurrency tests before, I meant adding randomized delays to operations that involve cross-thread coordination via atomics (or higher level primitives that invoke them internally). These might be OK as a benchmark so you can compare results on your own machine before and after changes, but I don't think they are useful as tests.

Reply:
Please try running the tests with master's `blocking.rs`. They will panic.

Reply:
They panic on this branch on a fairly modern machine, so I'm not sure they are useful on machines other than yours. In general, timing-based performance tests are an antipattern, as they have big reproducibility issues. If you want to do this kind of testing, it's better to use something like cachegrind to count CPU instructions executed.

Reply:
Because they don't actually assert on anything now (which was what caused them to fail on my laptop), I think they are better as benchmarks, right? I don't think it makes sense to have tests that can't fail.

Reply:
> I don't think it makes sense to have tests that can't fail.

Please read my comments carefully. You can see what it tests in action:
https://asciinema.org/a/R6b3UcFo0xfNo1sCPOPJrB04T

> when I said concurrency tests before, I meant adding randomized delays to operations that involve cross-thread coordination via atomics (or higher level primitives that invoke them internally).

Unexposed architecture shouldn't be tested with whitebox fashion.

Reply:
Where are the assertions that fail? The asciicinema is too long to watch. 

Possible concurrent interleavings are exposed architecture, and this kind of test has resulted in a huge number of bugs being discovered in various lock-free projects I've been involved in. Concurrency is really hard to get right sometimes, and the original PR with more unserialized atomic accesses would have had bugs jump out pretty quickly with this approach.